prog/dftb+/lib_dftb/encharges.F90

```diff
@@ -8,28 +8,442 @@
+
+    !> lattice vectors if periodic
+    real(dp) :: latVecs(3, 3)
+
+    !> Volume of the unit cell
+    real(dp) :: vol
+
+    !> evaluate Ewald parameter
+    logical :: tAutoEwald
+
+    !> Ewald tolerance
+    real(dp) :: tolEwald
+
+    !> Parameter for Ewald summation.
+    real(dp) :: parEwald
+
+    !> is this periodic
+    logical :: tPeriodic
+
+    !> Contains the points included in the reciprocal sum.
+    !> The set should not include the origin or inversion related points.
+    real(dp), allocatable :: recPoint(:, :)
+
+    !> are the coordinates current?
+    logical :: tCoordsUpdated
+
+    !> electrostatic energy
+    real(dp), allocatable :: energies(:)
+
+    !> force contributions
+    real(dp), allocatable :: gradients(:, :)
+
+    !> stress tensor
+    real(dp), allocatable :: stress(:, :)
+
+    !> partial charges
+    real(dp), allocatable :: charges(:)
+
+    !> derivative of partial charges w.r.t. coordinates
+    real(dp), allocatable :: dqdr(:, :, :)
+
+    !> derivative of partial charges w.r.t. strain deformations
+    real(dp), allocatable :: dqdL(:, :, :)
+
+  contains
+
+    !> initialize container from input
+    procedure :: initialize
+
+    !> update internal store of coordinates
+    procedure :: updateCoords
+
+    !> update internal store of lattice vectors
+    procedure :: updateLatVecs
+
+    !> return energy contribution
+    procedure :: getEnergies
+
+    !> return energy contribution
+    procedure :: getCharges
+
+    !> return force contribution
+    generic :: addGradients => addGradientsEnergy, addGradientsCharges
+
+    !> return force contribution for electrostatic energy
+    procedure, private :: addGradientsEnergy
+
+    !> return force contribution for charge derivatives
+    procedure, private :: addGradientsCharges
+
+    !> return stress tensor contribution
+    generic :: addStress => addStressEnergy, addStressCharges
+
+    !> return stress tensor contribution for electrostatic energy
+    procedure, private :: addStressEnergy
+
+    !> return stress tensor contribution for charge derivatives
+    procedure, private :: addStressCharges
+
+    !> cutoff distance in real space for EEQ
+    procedure :: getRCutoff
+
+  end type TEeqCont
+
+
 contains
 
+
+  subroutine initialize(this, input, withEnergies, withCharges, nAtom, latVecs)
+
+    !> Instance of EEQ container
+    class(TEeqCont), intent(out) :: this
+
+    !> Input for the EEQ model
+    type(eeqInput), intent(in) :: input
+
+    !> Enable electrostatic energy calculation + derivatives
+    logical, intent(in) :: withEnergies
+
+    !> Enable partial charge calculation + derivatives
+    logical, intent(in) :: withCharges
+
+    !> Nr. of atoms in the system.
+    integer, intent(in) :: nAtom
+
+    !> Lattice vectors, if the system is periodic.
+    real(dp), intent(in), optional :: latVecs(:, :)
+
+    real(dp) :: recVecs(3, 3), maxGEwald
+
+    this%tPeriodic = present(latVecs)
+
+    this%param = input%TEeqParam
```

Comment:
@bhourahine What about this part? It is using automatic allocation of LHS on purpose. Here the purpose is to make this part of the initialization less error prone.

Reply:
As that is the deliberate intention and its not in a loop, I don't see why. @aradi opinions ?